dev-portal/src/components/MessageList.jsx

```diff
@@ -0,0 +1,35 @@
+import React, { useState } from 'react'
+
+export const MessageList = ({ messages, dismissMessage, renderers }) => (
+  <>
+    {messages.map((message, index) => {
+      const { type, ...payload } = message
+      if (renderers[type]) {
+        return (
+          <React.Fragment key={index}>
+            {renderers[type](payload, () => dismissMessage(message))}
+          </React.Fragment>
+        )
+      }
+      throw new Error(`Unknown message type: ${type.toString()}`)
+    })}
+  </>
+)
+
+export const useMessageQueue = initialMessages => {
+  const [messages, setMessages] = useState(initialMessages || [])
+
+  const sendMessage = target => setMessages([...messages, target])
+  const dismissMessage = target => {
+    const deleteIndex = messages.findIndex(message => message === target)
+    if (deleteIndex === -1) {
+      throw new Error('Message not found')
+    }
+    setMessages([
+      ...messages.slice(0, deleteIndex),
+      ...messages.slice(deleteIndex + 1),
+    ])
```

Comment:
Yikes. Why not `.filter`?

Reply:
I only wanted for the first "matching" message to be removed from `messages`, and `.filter()` would unnecessarily check the entire array. But I think `.filter()` should be sufficient since `messages` will typically be small.

I'll change this to use `.filter()` instead.